src/ripple/app/tx/impl/DepositPreauth.cpp

```diff
@@ -0,0 +1,195 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/DepositPreauth.h>
+#include <ripple/basics/Log.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/st.h>
+#include <ripple/ledger/View.h>
+
+namespace ripple {
+
+TER
+DepositPreauth::preflight (PreflightContext const& ctx)
+{
+    if (! ctx.rules.enabled (featureDepositPreauth))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    std::uint32_t const uTxFlags = tx.getFlags ();
+
+    if (uTxFlags & tfUniversalMask)
+    {
+        JLOG(j.trace()) <<
+            "Malformed transaction: Invalid flags set.";
+        return temINVALID_FLAG;
+    }
+
+    auto const auth = ctx.tx[~sfAuthorize];
+    auto const unauth = ctx.tx[~sfUnauthorize];
+    if ((auth.is_initialized()) == unauth.is_initialized())
```

Comment:
Micronit: extra parens around `auth.is_initialized()` aren't needed.

Reply:
Note: std::optional doesn't have `is_initialized` and it's deprecated in boost. I wouldn't change this code though, we'll deal with it if we ever switch to `std::optional`.